### PR TITLE
Add Input Action and NodePath completion

### PR DESCRIPTION
**Pull request by jsbeckr**
I implemented input action and `NodePath` completion by using https://github.com/godotengine/godot/tree/master/modules/mono/editor/GodotTools/GodotTools.IdeMessaging (it's [MIT](https://github.com/godotengine/godot/blob/master/LICENSE.txt)). I had to copy the code from the Godot repo, because the nuget version is outdated. But that would be just a temporary thing, because Godot plans to publish proper nuget packages for communicating with the Godot IDE in the future. Once that's done we would integrate them and remove the `GodotTools.IdeMessaging` code and assembly from the plugin.

### TODO

- [x] Account for script placement in scene tree
- [x] Indicator about the status of the connection to the Godot IDE

### Possible Improvements:

- Refactor `ResourcePath` completion to use the `IdeMessaging` as well
- (maybe) Implement `ScenePaths` completion
- (maybe) Implement `PlayRequest`s / `OpenFileRequest`s via `IdeMessaging`

### NodePath Completion

https://github.com/JetBrains/godot-support/assets/147412/43acb95a-10e1-4018-911e-298606918a25

### Input Action Completion

https://github.com/JetBrains/godot-support/assets/147412/1f51cfa9-4693-4945-ba9b-a20d44e49bf1

P.S.: I signed the contributor agreement.

### CODE REVIEW

**Comment by jsbeckr** [comment]
Closes #86

**Comment by jsbeckr** [comment]
NodePath completion should work now exactly how it works in VS Code and Godot Editor. It queries against the currently open scene and accounts for relative script location in the scene tree.

**Comment by van800** [comment]
Thank you! I will take a look shortly.
Mentioning @gipsond as he might have considerations.

**Comment by van800** [comment]
What are options for adding automated tests for this completion?
It can only be tested with GodotEditor running, right?

**Comment by jsbeckr** [comment]
> What are options for adding automated tests for this completion? It can only be tested with GodotEditor running, right?

I was thinking about that as well. And you are correct the Godot Editor has to run in order to accept connections from Rider. We could mock the Editor part but that would only accomplish testing the mock implementation imho. The only part that would be interesting to test would be that the correct methods are called in `GodotMessagingClient` depending on the caret position in a C# script. But tbh I have no clue how to implement that... I would love to see a bit more love for the Rider/Resharper SDK documentation in general :D

File: resharper/src/CSharp/Completions/GodotSmartCompletion.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            itemsCollected |= LookupInputActions(invocationExpression, context, collector);
+            itemsCollected |= LookupNodePaths(invocationExpression, context, collector);
+
+            return itemsCollected;
+        }
+
+        private static bool LookupNodePaths(IInvocationExpression invocationExpression, CSharpCodeCompletionContext context,
+            IItemsCollector collector)
+        {
+            if (NodePathMethods.Methods.Contains(invocationExpression.InvokedMethodName()))
```
Comment (van800) [inline resharper/src/CSharp/Completions/GodotSmartCompletion.cs:60]: Why we are not checking the type here? I would expect it to be `Node` or its descendants.

Comment (jsbeckr) [reply resharper/src/CSharp/Completions/GodotSmartCompletion.cs]: You are right. I thought I had to implement a check for every Sub-Node (like `PanelContainer` etc.). But it works just with `Node`. Added the check.

**Comment by van800** [comment]
> > What are options for adding automated tests for this completion? It can only be tested with GodotEditor running, right?
> 
> I was thinking about that as well. And you are correct the Godot Editor has to run in order to accept connections from Rider. We could mock the Editor part but that would only accomplish testing the mock implementation imho. The only part that would be interesting to test would be that the correct methods are called in `GodotMessagingClient` depending on the caret position in a C# script. But tbh I have no clue how to implement that... I would love to see a bit more love for the Rider/Resharper SDK documentation in general :D

I would be happy even without those tests for now. I just wanted to clarify.

File: rider/testData/solutions/ResCompletionTest/.godot/editor/NewScript.cs-folding-c4b8f5e741d04006ec5d8d0c099949b5.cfg
Comment (van800) [inline rider/testData/solutions/ResCompletionTest/.godot/editor/NewScript.cs-folding-c4b8f5e741d04006ec5d8d0c099949b5.cfg:1]: Why update the testdata? Can we revert those changes?

File: resharper/src/CSharp/Completions/InputActionMethods.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+using System.Collections.Generic;
+
+namespace JetBrains.ReSharper.Plugins.Godot.CSharp.Completions
+{
+    internal static class InputActionMethods
+    {
+        internal static readonly List<string> Methods = new()
+        {
+            "ActionPress", "ActionRelease", "GetActionRawStrength", "GetActionStrength", "GetAxis",
+            "IsActionJustPressed", "IsActionJustReleased", "IsActionPressed"
```
Comment (van800) [inline resharper/src/CSharp/Completions/InputActionMethods.cs:10]: Can we drop the check over hard-coded list of methods? And check for the `StringName` type of the parameter instead?

**Review by van800 — CHANGES_REQUESTED** [review]

File: rider/testData/solutions/ResCompletionTest/.godot/editor/NewScript.cs-folding-c4b8f5e741d04006ec5d8d0c099949b5.cfg
Comment (jsbeckr) [reply rider/testData/solutions/ResCompletionTest/.godot/editor/NewScript.cs-folding-c4b8f5e741d04006ec5d8d0c099949b5.cfg:1]: Yeah we can revert those. I was just testing with the included Godot project in the beginning.

File: resharper/src/CSharp/Completions/InputActionMethods.cs (hunk 2 above)
Comment (jsbeckr) [reply resharper/src/CSharp/Completions/InputActionMethods.cs:10]: Sounds good. I will take a look at that in the evening 👍

**Comment by van800** [comment]
I have added the connection indicator to the UI. Would be nice if you check on your side, that it works fine.
I have a suspicion, that it doens't fire, when the GodotEditor process is killed. But maybe it is not a big deal.

**Comment by jsbeckr** [comment]
> I have added the connection indicator to the UI. Would be nice if you check on your side, that it works fine. I have a suspicion, that it doens't fire, when the GodotEditor process is killed. But maybe it is not a big deal.

So I tried it on MacOS and Windows and it works perfectly. Even when killing the Godot editor process. Nice touch with the disconnected icon btw. :D

File: resharper/src/CSharp/Completions/InputActionMethods.cs (hunk 2 above)
Comment (jsbeckr) [reply resharper/src/CSharp/Completions/InputActionMethods.cs:10]: Okay... I tried my luck for a few hours. But I'm can't wrap my head around traversing the syntax tree. But I think it should be doable. We could even check for the type of `ILocalVariableDeclaration` and if it is a `NodePath` we could support auto complete for something like this:

```CSharp
NodePath test = "Icon"
```

Could give me a hint of how to implement that?

**Comment by van800** [comment]
Would this hang forever in case, if Godot Editor is refusing the connection?
https://github.com/JetBrains/godot-support/blob/235ee7fed83167ab2d6dc6c9ee21801c690d4157/resharper/GodotTools.IdeMessaging/Client.cs#L293-L294

File: resharper/src/CSharp/Completions/InputActionMethods.cs (hunk 2 above)
Comment (van800) [reply resharper/src/CSharp/Completions/InputActionMethods.cs]: I had to do some trial-error, please see https://github.com/JetBrains/godot-support/pull/102/commits/2a663deb8ae7b3d1db8dcd4840c610afb740841c

Comment (van800) [reply resharper/src/CSharp/Completions/InputActionMethods.cs]: Would you do the same for `LookupNodePaths`?

Comment (jsbeckr) [reply resharper/src/CSharp/Completions/InputActionMethods.cs]: Uff easier than I thought! I added support for `NodePath` variable declarations as well.

**Comment by Perndoe** [comment]
Wow, this is so cool

**Comment by van800** [comment]
@jsbeckr Is there anything else, which you want to accomplish within this PR?

**Comment by jsbeckr** [comment]
> @jsbeckr Is there anything else, which you want to accomplish within this PR?

Nope! I am super happy with this! 👍

**Comment by Perndoe** [comment]
> Closes #86

👍

**Comment by van800** [comment]
@jsbeckr some of your commits were from a different git user, I have fixed that and force-pushed.

**Comment by van800** [comment]
Plugin is published - 2023.2.121.

**Comment by geowarin** [comment]
Testing this with the master version of godot (rider EAP 6) has mixed result:

Sometines, When I open a C# script from Godot, it seems that rider can't connect to godot and I get a null reference when asking for a completion (annot reproduce rigth now, I will send the stacktrace when I get a new one)

This error goes away if I restart rider.

Then, Input completions work great :+1: 

However, for NodePath completions I get errors in the Godot console:

```
Caller thread can't call this function in this node (/root/@EditorNode@17639/@Control@697/@Panel@698/@VBoxContainer@706/@HSplitContainer@709/@HSplitContainer@717/@HSplitContainer@725/@VBoxContainer@726/@VSplitContainer@728/@VSplitContainer@753/@VBoxContainer@754/@PanelContainer@799/MainScreen/@CanvasItemEditor@10089/@VSplitContainer@9914/@HSplitContainer@9916/@HSplitContainer@9918/@Control@9919/@SubViewportContainer@9920/@SubViewport@9921/TrollSlasher). Use call_deferred() or call_thread_group() instead.
```

This is probably linked to the new Threading model in godot 4.1 (see [godot#78149](https://github.com/godotengine/godot/issues/78149))

Did any of you test this with the 4.1 RC version of Godot?

**Comment by van800** [comment]
I was using Godot 4.0.3

**Comment by jsbeckr** [comment]
The icon is in the top right next to the build and run icon. Maybe it's only visible with Next UI enabled? 

The exception that you see should be an internal problem of Godot. We just connect via TCP and send a message. I'm not not at home right now so it's hard to test for me. Would be a shame if it's not working for 4.1.

**Comment by geowarin** [comment]
Yes I see the icon, I was expecting it to be on the bottom panel for some reason :smile: 

Might be a godot bug, new threading model has been causing a few issues here and there, maybe the interaction with the IDE messaging has been undetected before now.

**Comment by van800** [comment]
The connection indicator should work in both new/old UI. I have tried that.

**Comment by jsbeckr** [comment]
I guess we have to open a bug for that. Indeed NodePath completion does not work in 4.1 right now. Curiously enough it's not Godot (I think). During development of this feature I build the plugin to use it in my current Rider 2023.1.3 installation.

With that version I can still get NodePath completion. So in my opinion we introduced the problem during the refactoring (but that only affects 4.1 for whatever reason) or it has something to do with the Rider EAP 2023.2 in combination with Godot 4.1. Which also sounds weird. ¯\_(ツ)_/¯

**Comment by van800** [comment]
Tried it today again with Godot 4.2.1 and Rider 2023.3.3. It works fine. Make sure, GodotEditor is running and eyes of the Godot icon in the Rider toolbar is opened.
image

**Pull request merged**